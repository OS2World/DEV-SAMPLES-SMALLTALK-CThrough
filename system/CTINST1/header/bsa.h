#ifndef BSA_INCLUDED
#define BSA_INCLUDED

/*
 *		This is the standard BSA preamble header
 */


//------------------------------ standard global declarations
typedef unsigned int	size_t;
typedef unsigned long	ulong;

extern void*	operator new (size_t);
extern void		operator delete (void*);


//------------------------------ CT HeapManager Section
#ifdef __HEAPTRACE__

//--- function prototypes required by macros
extern void*	operator new (size_t,int);
extern void*	wrapNew (void*);
extern int		beginNew (char*,int,char*,size_t,unsigned,unsigned);
extern int		beginDelete (char*,int,void*,int);
extern void		endDelete (char*,int,void*,int);
extern char*	strndupDbg (char*,int,char*,unsigned long);

//--- macros for Heap operations
#define NEW(type,par)			 (type*)( wrapNew( beginNew(__FILE__, __LINE__, #type, sizeof(type), 0, 0 ) ? \
										new(1) type par : 0) )
#define NEW_VEC(type,dim)		 (type*)( wrapNew( beginNew(__FILE__, __LINE__, #type, sizeof(type), dim, 0) ? \
										new(1) type [dim] : 0 ) )
#define NEW_VEC2(type,dim1,dim2) (type*)( wrapNew( beginNew(__FILE__, __LINE__, #type, sizeof(type), dim1, dim2) ? \
										new(1) type [dim1][dim2] : 0 ) )

#define NEW_STR(pchar)			strndupDbg(__FILE__, __LINE__, pchar, 0)
#define NEW_STRN(pchar,len)		strndupDbg(__FILE__,__LINE__, pchar, len)


#define DELETE(adr)			if(adr){ beginDelete( __FILE__, __LINE__, adr, 0); \
										delete adr; endDelete( __FILE__, __LINE__, adr, 0); adr = 0;} else {;}
#define DELETE_VEC(adr)		if(adr){ beginDelete( __FILE__, __LINE__, adr, 1); \
										delete [] adr; endDelete( __FILE__, __LINE__, adr, 1); adr = 0; } else {;}
#define DELETE_NZ(adr)		if(1){ beginDelete( __FILE__, __LINE__, adr, 0); \
										delete [] adr; endDelete( __FILE__, __LINE__, adr, 0); } else {;}
#define DELETE_STR(adr)		if(adr){ beginDelete( __FILE__, __LINE__, adr, 1); \
										delete [] adr; endDelete( __FILE__, __LINE__, adr, 1); adr = 0; } else {;}

//--- non heaptrace macro versions
#else

//--- function prototypes required by macros
extern char* strndup (char*,unsigned long);

//--- macros for Heap operations
#define NEW(type,par)				new type par
#define NEW_VEC(type,dim)			new type [dim]
#define NEW_VEC2(type,dim1,dim2)	new type [dim1] [dim2]

#define NEW_STR(pchar)				strndup( pchar, 0)
#define NEW_STRN(pchar,len)			strndup( pchar, len)

#define DELETE(adr)					if(adr){ delete adr; adr=0; } else {;}
#define DELETE_VEC(adr)				if(adr){ delete [] adr; adr=0; } else {;}
#define DELETE_STR(adr)				if(adr){ delete [] adr; adr=0; } else {;}
#define DELETE_NZ(adr)				delete adr;

#endif //__HEAPTRACE__



#endif // BSA_INCLUDED