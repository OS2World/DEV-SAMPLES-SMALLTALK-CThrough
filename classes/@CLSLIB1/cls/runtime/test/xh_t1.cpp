#ifndef INC_XH_T1
#define INC_XH_T1

#ifndef __FIRST__
#define __FIRST__
#define __IMPL__XH_T1
#endif

/////V XH_t1 PCM f:\cls_ibm\cls\runtime\test 2  PM 06.04.95 15:05:50
/*
/////H
18.12.94 01:57 PM 2 copied from: XH_t1 TOS f:\cls_ibm\cls\runtime\test 2  PM 19.10.94 01:09:49
30.03.95 10:34 PM 0 archived: XH_t1 PCM f:\cls_ibm\cls\runtime\test 0 PM 18.12.94 01:57:50
/////
*/

/////1
#undef inline

#include <bsa.h>

/////I conio.h @ @ @ @ pre 
#include <conio.h>

/////I stdio.h @ @ @ @ pre 
#include <stdio.h>

/////I XH @ @ @ @ class pre 
#include <XH.cpp>

/////T Aux pre 
struct X {
	char	*msg;

	X ( char* s )
		{
			msg = s;
			printf( "\t\t*** X (\"%s\") object created\n", msg);
		}

	~X ()
		{
			printf( "\t\t*** X (\"%s\") object destroyed\n", msg);
		}
};


class IllegalValue {};
/////


#ifndef __INLINE__
#define inline
#endif

/////C XH_t1 @ @ app define:_VIO_ 
class XH_t1

{

public:
	static int  doAccessViolation ();
	static int  doCppException ();
	static int  doSomeExceptions (char);
	static int  doUserRaise ();
	static int  doZeroDiv ();
	static int  testInt (int);
};

	int _System  secondThread (ULONG);

/////2
#undef inline


#if (defined __INLINE__) || (defined __IMPL__XH_T1)

#ifndef __INLINE__
#define inline
#endif

/////
#endif

/////3
#undef inline

#ifdef __IMPL__XH_T1
/////F doAccessViolation @ @ class public 
int XH_t1:: doAccessViolation ()
{
	int  a;
	char *c = 0;

	puts( "      entering func doAccessViolation()");

	puts( "\t\t*** now we force an access violation");
	a = *c;

	puts( "      leaving func doAccessViolation()");
	return a;
}

/////F doCppException @ @ class public 
int XH_t1:: doCppException ()
{
	puts( "      entering func doCppException()");
	int 	a = 0;
	X		x1("doCppException");				// unwinded by RAISE -> retry

	/*
	 * This function demonstrates how to combine C++ exception handling (try,
	 * catch, throw) with the CTHROUGH exception handling (SAVEPOINT, RAISE).
	 * C++ exceptions cannot handle OS/2 raised exceptions, but are capable of
	 * storing some additional information about the exception in the
	 * corresponding exception class object. Another difference is the control
	 * flow: C++ exceptions cause a "forward" jump to the next catch
	 * block, CTHROUGH exceptions do a "fall back" to the last SAVEPOINT.
	 * While the first one is more structured (in order to "repair" the
	 * cause of an exception), the last one is more suitable for GUI programs
	 * (with an appropriate "restart" point)
	 */

	try {
		X	x2("doCppException try {..x..}");		// unwinded by throw in testInt()

		testInt( a);							// will throw an IllegalValue exception
		printf( "\t\t*** 10 / a = %d\n", 10 / a);
	}
	catch ( IllegalValue ) {
		puts( "\t\t*** caught an \"IllegalValue\" Cpp exception in doCppException()");

		puts( "\t\t*** we now RAISE a corresponding user exception");
		XStream << "IllegalValue condition detected";
		XRAISE( DA_RETRY);
	}

	puts( "      leaving func doCppException()");
	return 1;
}

/////F doSomeExceptions @ @ class public 
int XH_t1:: doSomeExceptions ( char cmd )
{
	printf ( "   entering func doSomeExceptions( %c )\n", cmd);

	switch ( cmd ) {
		case '1' :	XH_t1::doZeroDiv(); 		break;
		case '2' :	XH_t1::doUserRaise(); 		break;
		case '3' :	XH_t1::doAccessViolation(); break;
		case '4' :	XH_t1::doCppException();	break;
	}

	puts( "   leaving func doSomeExceptions()");

	return 1;
}

/////F doUserRaise @ @ class public 
int XH_t1:: doUserRaise ()
{
	puts( "      entering func doUserRaise()");

	SET_SAVEPOINT( sp);		// set savepoint for RETRY respond
	puts( "\t\t*** we just passed the SAVPOINT in doUserRaise()");


	puts( "\t\t*** we now RAISE a user exception");
	XStream << "I got a problem here";	// raise user exception ...
	XRAISE( DA_RETRY);					// with default action RETRY

	puts( "      leaving func doUserRaise()");
	return 1;
}

/////F doZeroDiv @ @ class public 
int XH_t1:: doZeroDiv ()
{
	puts( "      entering func doZeroDiv()");
	int a=10,b=0;
	X	x("doZeroDiv");	 // will be automatically unwinded if user respond = "retry"

	puts( "\t\t*** now we force a division by zero");
	a = a/b;

	puts( "      leaving func doZeroDiv()");
	return a;
}

/////
#ifdef __APPCLASS__
/////F main @ @ global 
int main ( int argc, char* argv[] )
{
	/*
	 *	*** exception handling demo app ***
	 *	
	 *	Importing XH will cause a duplex pipe connection to the external XProcessor
	 *	XHProc.exe ( if running).
	 *	Threads are linked to XH processing by macro REGISTER_XCPT.
	 *	XH supports internal OS/2 exceptions ( like zero devision) as well as user raised ones.
	 *	For user exception simply write a small text explaining your runtime problem to the global
	 *	XH instance 'XStream' and raise it with a default action ( DA_ABORT, DA_RETRY, DA_IGNORE).
	 *	If XHProc.exe is running, it pops up an error-desribing MessageBox ( or VIO pane, depending
	 *	of the raising app type). This respond box provides 3 possible reactions ( ABORT, RETRY, IGNORE).
	 *	OS/2 exceptions are suggested ABORT by default, for possible serious runtime problems.
	 *	The user respond is passed back to XH ( queued threads are internally dispatched meanwhile).
	 *	ABORT will terminate the using app, RETRY jumps to the last registered SAVEPOINT
	 *	of the raising thread ( fixed by macro SET_SAVEPOINT(sp) ) and IGNORE skips the exception point.
	 *	If the XProcessor is not running the X-causing app will be terminated.
	 *	The raise condition is logged to the file 'XH.xlg' anyway. This file can be analyzed by BsaController's
	 *	Tools XLogViewer, that shows the raise condition of each entry, the describing text( zero division,
	 *	user raise ...) as well as it's call stack back to the X-root.
	 */

	TID tid;
	REGISTER_XCPT;

	//--- we do some IO that requires sync, so reset buffers here
	setbuf( stdin, 0);
	setbuf( stdout, 0);

	puts( "!! please make sure that the XProcessor has been started");
	puts( "!! prior to XH_t1 in order to get a visual feedback on the");
	puts( "!! exceptions. Otherwise XH_t1 will die gracefully when");
	puts( "!! encountering the first exception (but will leave an exception");
	puts( "!! log file XCPT.XLG that can be examined later with the XLogViewer\n");

	//--- start another thread to demonstrate usage in multi thread environments
	DosCreateThread( &tid, (PFNTHREAD) secondThread, 0, 0, 65535);

	DosWaitThread( &tid, 0);
	UNREGISTER_XCPT;
	return 0;
}
/////
#endif

/////F secondThread @ @ global 
int _System secondThread( ULONG args)
{
	puts( "entering thread func secondThread()");
	char	c;

	REGISTER_XCPT;			// link thread to exception handler
	SET_SAVEPOINT( sp);		// set savepoint for RETRY respond
	puts( "\t\t*** we just passed the SAVPOINT in secondThread()");

	/*
	 * The RETRY option of the Exception handling is only useful if the
	 * application has a suitable "restart point" (=SAVEPOINT). This is
	 * usually the case with GUI apps (there always is some kind of a
	 * "dispatcher loop"). In this example, we simulate this behaviour with
	 * the following loop
	 */

	while ( 1 ) {
		printf( ">>enter command ('1'..'4', other key for exit) : ");
		if ( (c = getche()) > 52 || c < 49 ){
			puts( "\nloop terminated by user");
			break;
		}

		puts( "\n\t\t*** next loop in secondThread()");
		XH_t1::doSomeExceptions( c);
	}

	puts( "leaving thread func secondThread()");

	UNREGISTER_XCPT;		// remove thread from X handler
	return 1;
}

/////F testInt @ @ class public 
int XH_t1:: testInt ( int a )
{
	puts( "         entering func testInt()");
	X	x("testInt");			// will be automatically unwinded after throw

	if ( a == 0 ) {
		puts( "\t\t*** we now throw an \"IllegalValue\" exception from testInt()");
		throw IllegalValue();
	}

	puts( "         leaving func testInt()");
	return 1;
}

/////
#endif

#endif
