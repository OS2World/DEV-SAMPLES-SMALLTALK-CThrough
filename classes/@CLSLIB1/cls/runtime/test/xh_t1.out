/////V xh_t1  f:\cls_ibm\cls\runtime\test 1  PM 05.09.94 21:54:16

/////I XH_T1 exception handling example @ 0 @  @ 
This example consists of a single file that is used to demonstrate
the CTHROUGH exception handling class XH.

Exceptions can be raised either direct (so called "user exceptions", 
e.g. in "doUserRaise()") or indirect via a OS/2 exception handler.

If the XProcessor process was started (e.g. from the CThrough
folder), the user will see an appropriate exception message and
will be asked for further directions ("abort, retry, ignore").

If the XProcessor was not started, the application will be aborted
on the first exception.

In both cases, an Exception log file (XCPT.XLG), residing in the
applications directory,  will be updated.

This logfile can be analyzed by the XLogViewer that can be started
from the Controller.

/////I The XProcessor @ 1 @  @ 
The XProcessor is an external program that is represented
by a small StdWindow (which usually is minimized).

It listens to a NamedPipe that is used to report exceptions. If such
a message is received, a MessageBox pops up, asking the user
what to do with the ill behaving application (abort, retry, ignore).
This message box actually belongs to the XProcessor. If it does
not get active because of PM problems in the ill behaving app,
the XProcessor is able to detect and kill such PM user lockouts.

The XProcessor has to be started just once. It can serve multiple
simultaneously running applications.

The XProcessor is a small, simple app that can be tailored to
specific needs (e.g. to report errors to a sysop on networked
machines).

/////I The XLog @ 1 @  @ 
There is an Exception log file XLOG.XLG written either by the
XProcessor or by the ill behaving app itself. This file is located
in the same directory as the application itself.

The XLog has one entry for every exception that occurs. All
Entries contain info about 

	date
	time
	module name and process id of ill behaving process
	(optional) exception message
	call stack for exception

/////I The XLogViewer @ 1 @  @ 
The XLogViewer is a tool that can be used within CTHROUGH. It
displays the contents of a XLog file and enables the user to
further analyze exceptions that are stored in the XLog.

If the XLogViewer is able to locate a (converted) linker map file
for the application (*.mp, automatically created if the "map"
option was set in the AppsBuilder), it will load this mapfile in
order to display the call stack in its lower pane (the StackPane).

From this pane, the user can open a ClassEditor that will be
positioned automatically on the appropriate function.

If the mapfile is younger than the app or the exception entry,
the XLogViewer will mark the entry as "outdated" (grayed).
The call stack of outdated entries can not be analyzed (no
corresponding mapfile).

/////I SavePoints @ 1 @  @ 
The XH class uses a "SavePoint" as an abstraction of the 
setjmp/longjmp mechanism.

SavePoints can be set at places that form suitable restart points
of a program (e.g. in front of dispatcher loops).

When an exception occurs and the user selects the "retry" option
from the Exception MessageBox (launched by the XProcessor), the
program will do a longjmp to the last SavePoint.

>> This is done in combination with a stack unwind <<

There are several examples of SavePoints in XH_T1
(e.g. in function "secondThread()")

/////I explicit raise @ 1 @  @ 
Explicit "user" raises can be done with the 

		XRAISE( defaultAction )

macro. Prior to the execution of this macro, there can be text
written to the Exception stream (that will appear as message
text in the XLog or the Exception MessageBox).

/////I multile threads @ 1 @  @ 
XH can be used in a multi thread application. In order to enable
the handling of OS/2 exceptions (access violations etc), every
thread function has to use the REGISTER_XCPT and
UNREGISTER_XCPT macros (see function "secondThread()").

/////I ignore an exception @ 1 @  @ 
Even if there is a "ignore" respond in MessageBoxes on OS/2
exceptions (e.g. access violations and division by zero), this may
not be useful in many cases. Ignoring such an exception most times
will result just in a follow on exception. However, the program
should continue with the next instruction if this option was selected.

/////I C++ and CTHROUGH exception handling @ 1 @  @ 
There are some differences between the C++ and the CTHROUGH
exception handling.

Function "doCppException()" demonstrates how both methods can
be combined.

C++ mainly uses three constructs to handle exceptions:

	try { .. }
	catch ( <exceptionClass> ) { .. }
	throw <exceptionClass object>

System exceptions like access violations or division by zeros
cannot be handled directly via catch blocks. For every exception
there must be an explicit raise.

C++ always does a "jump forward" from the raise to the matching
catch block, "retry" functionality must be programmed explicitely
(by means of a local loop or a setjmp/longjmp).

Generally spoken, the C++ way to do things is more appropriate to
handle "high level" exceptions that are detected before they can
cause "low level" exceptions like access violations etc.


CTHROUGH exception handling uses 2 constructs (macros):

	SAVEPOINT
	RAISE

It can be used for implicitely "raised" OS/2 exceptions (division by zero
etc.) and for explicitely raised (by means of the RAISE macro) - so
called - user exceptions.

All exceptions are handled in the same way, there is no thing like an
exception object that can store additional info. If the XProcessor is
started, the user has the choice to:

	abort the process
	retry the operation (jump back to last SAVEPOINT)
	ignore the exception (continue with the next instruction)

If the user respond is a "retry" there will be a "fall back" to the
last previous SAVEPOINT. This also incorporates a stack unwind.

In addition, there can be text output that will be attached to the
exception in the XProcessor MessageBox as well as the XLog file.
Since this output is written to a pipe (and processed in the XProcessor),
there is no restriction about the size of the text (could be used to
"dump" objects).
